app/assets/javascripts/components/generic_object/generic-object-definition-treeview-component.js

```diff
@@ -0,0 +1,168 @@
+      vm.setActiveNode();
+    });
+  };
+
+  vm.nodeSelect = function(node) {
+    var key = node.key.split('_');
+
+    miqService.sparkleOn();
+    switch (key[0]) {
+      case 'god':
+        if (key[1] === 'root') {
+          $window.location.href = vm.showListUrl;
+        } else {
+          $window.location.href = vm.showUrl + '/' + key[1];
+        }
+        break;
+      case 'cbs':
+        $window.location.href = vm.showUrl + '/' + node.parentNodeId + '?cbs=' + key[1];
+        break;
+      case 'cb':
+        $window.location.href = vm.showUrl + '/' + node.parentNodeId + '?cb=' + key[1];
+        break;
+      case 'actions':
+        $window.location.href = vm.showUrl + '/' + node.parentNodeId + '?actions=true';
+        break;
+      default:
+        $window.location.href = vm.showListUrl;
+    }
+  };
+
+  vm.setActiveNode = function() {
+    var activeNode;
+
+    var urlParams = new URLSearchParams($window.location.search);
+
+    if (urlParams.get('cbs')) {
+      activeNode = 'cbs_' + urlParams.get('cbs');
+    } else if (urlParams.get('cb')) {
+      activeNode = 'cb_' + urlParams.get('cb');
+    } else if (urlParams.get('actions')) {
+      activeNode = 'actions_' + $window.location.href.split("/").pop().split('?')[0];
+    } else if ($window.location.href.includes(vm.showListUrl)) {
+      activeNode = 'god_root';
+    } else if ($window.location.href.includes(vm.showUrl)) {
+      activeNode = 'god_' + $window.location.href.split("/").pop();
+    }
+
+    vm.activeNode =  { key: activeNode };
+  };
+
+  vm.$onDestroy = function() {
+    $window.top.name = '';
+  };
+
+  // private functions
+
+  function setGenericObjectDefinitionNodes() {
+    API.get('/api/generic_object_definitions?expand=resources&attributes=name,picture.image_href,custom_button_sets,custom_buttons&sort_by=name&sort_options=ignore_case&sort_order=asc')
+      .then(setNodes)
+      .catch(miqService.handleFailure);
+  }
+
+  function setNodes(response) {
+    var image;
+    var icon;
+    var godNodes = [];
+    var treeDataObj = [];
+
+    _.forEach(response.resources, function(resource) {
+      if (resource.picture && resource.picture.image_href) {
+        image = resource.picture.image_href;
+        icon = undefined;
+      } else {
+        image = undefined;
+        icon = 'fa fa-file-text-o';
+      }
+
+      godNodes.push({
+        key: 'god_' + resource.id,
+        text: resource.name,
+        tooltip: resource.name,
+        image: image,
+        icon: icon,
+        state: {expanded: false},
+        nodes: createCustomButtonAndSetNodes(resource.custom_button_sets, resource.custom_buttons, resource.id),
+      });
+    });
+
+    treeDataObj = [{
+      key: 'god_root',
+      text: __('All Generic Object Classes'),
+      tooltip: __('All Generic Object Classes'),
+      icon: 'pficon pficon-folder-close',
+      state: {expanded: true},
+      nodes: godNodes,
+    }];
+
+    vm.treeData = JSON.stringify(treeDataObj);
```

Comment:
This is wrong, you should not build nodes in the JS, they should be created in a `TreeBuilder` subclass as every other tree.

Reply:
Please explain why since I do not see a problem with this.

Using `TreeBuilder` effectively means saying NO to using REST API. 
Why would we want that when we can use REST API?

Reply:
@AparnaKarve because it is inconsistent, introduces a lot of exceptions and technical debt. Last time we were chatting about this, I told you that we're starting to build an API-like [controller for the trees](https://github.com/skateman/manageiq-ui-classic/blob/6d51689de779ab9f0a5a26b6b8af8d6ef2f53c24/app/controllers/tree_controller.rb) that will eventually land in the REST API.

Reply:
@AparnaKarve : how many API calls do you need to build the tree? Can you estimate that number based on the number of generic objects or provide any other metrics that will show us how this will scale?

I thought we had an agreement that we want to build the tree on the server side.

Also I thought we had an agreement that we want to find a common pattern for all the trees and convert all the trees to use the same API.

It may make sense to use client-side tree building if
 1) it will not send too many API requests -- therefor I am asking above.
 2) this tree is going to be special in some way so that it would not make sense to share the common pattern

Thx!

Reply:
@martinpovolny One API only, and that's it!

Everything that I need can be accomplished with one single API call, and that is precisely why I went with the REST API approach.

Reply:
well then it's pretty neat

Reply:
How many generic object will it be able to handle? A single API request, that's nice, but will it not kill the client when there are 100.000 or them?

Reply:
@skateman, @AparnaKarve : I am trying to uderstand why the pattern that @skateman suggested (and that I believe was meant to be generic for all the trees) was not used here.

A am not sure if we would be able to build the other trees in a similar way on the client. What do you think?

Generally I'd prefer to limit the number of different ways to do a single thing for sake of maintainability.